### ISSUE TITLE
Add possibility to set default attribute and element values

### ISSUE THREAD

**Issue by cmutel** [sami-m-g/pyecospold#25]
The current system of creating property methods in the `DataHelper` which come with built-in validation is quite nice, though some better documentation would help newcomers :) However, one thing we absolutely need is a way to specify default values for elements. This would make data entry much easier and user-friendly.

The ideal system for me would be to keep the `create_attribute` functionality as it is (i.e. we don't create `foo.x` with a default value on element creation), but only substitute in the default values when serializing. @sami-m-g I would appreciate some help with software architecture here, as all the `DataHelper` methods are static - which is good! - I don't think we can add default values in `DataHelper.create_attribute` without really ugly hacks.

Maybe one option would be to create a dictionary `__defaults` for each element class, e.g. 

```python
class Validation(etree.ElementBase):
    __defaults = {
        "proofReadingDetails": "Done to the best of the author's ability, but not well enough to override a default value",
        "otherDetails": "None provided"
    }
```

Then all serialization would require recursive traversal of all element class to substitute in default values where no attribute values are present. I don't love this but don't see anything better.

Another complication is that some defaults are programmatic, e.g. the default `Validation.proofReadingValidator` would be the first (or last?) author listed in another element. So maybe we just need a `finalize_for_serialization` function on each element class? This would give us flexibility.

**Comment by sami-m-g** [comment]
@cmutel, so you mean we only fill in default values when writing the final XML file? What value does this provide over just filling in the values on element creation?

Furthermore, we do have a current functionality for default values. Currently it works on reading only, but we should simply build on that to add writing capability.

Finally, for programmatic defaults, unless it's a small set of pre-defined options, I would rather simply let the users use the `Defaults` class for setting their own rules. We may decide on adding a couple more helper methods, maybe utilize `lxml.xpath` as well.

**Comment by cmutel** [comment]
I don't have a preference as to *when* it happens, or even how, but something like 20% of Ecospold 1 can always be default, and maybe another 10% can be defined in a config file (e.g. author details). The ideal system would be simple and transparent, so if we need to provide a `Defaults` class, it should be easy to read and understand, and be used by default.

@sami-m-g Maybe you could write a quick prototype in pseudocode?

**Comment by sami-m-g** [comment]
@cmutel
How about something like this:
`class DefaultableElementBase(etree.ElementBase):
    def _init(self) -> None:
        # Iterate over class properties and set defaults for Nones
        # Using Defaults._property_name_ or Defaults[_class_name_]._property_name_`

**Comment by cmutel** [comment]
@sami-m-g Sounds good. Here are some reasonable defaults:

* `dataset.timestamp`: lambda function for now
* `dataset.validCompanyCodes`: "CompanyCodes.xml"
* `dataset.validRegionalCodes`: "RegionalCodes.xml" 
* `dataset.validCategories`: "Categories.xml"
* `dataset.validUnits`: "Units.xml"
* `dataset.generator`: "pyecospold" with version set dynamically
* `referenceFunction.localName`: Copy of `referenceFunction.name`
* `referenceFunction.infrastructureProcess`: `True`
* `dataSetInformation.impactAssessmentResult`: `False`
* `dataSetInformation.timestamp`: lambda function for now
* `dataSetInformation.version`: `1.0`
* `dataSetInformation.internalVersion`: `1.0`

Not sure if this is covered elsewhere, but we also need to be set some element values based on the state of the dataset, e.g. `referenceFunction.datasetRelatesToProduct`, `referenceFunction.amount`, and `referenceFunction.unit` can be determined by iterating through the exchanges, `administrativeInformation.dataEntryBy` can come from a config file. So maybe it is cleaner to have a `prepare_for_serialization` method where we can set defaults and these computed values?

**Comment by sami-m-g** [comment]
@cmutel 
For `generator: "pyecospold" with version set dynamically`, do you mean our library version or the schema version as in 1 or 2? Can you kindly elaborate?

For `foo.timestamp: lambda function for now`, do we really want to handle the hassle of handling lambdas and strings as different config values? I assume the default for timestamp won't change frequently -otherwise it's not that much of a "default"- so the user would better go for something like:
```
# Include whatever imports needed
# Set value for default_timestamp
# Set the Defaults.property_name or Defaults[class_name].property_name to default_timestamp
````
What do you think?

For `referenceFunction.localName: Copy of referenceFunction.name` I would have suggested something but since you mention that "some element values based on the state of the dataset" maybe we can simply actually do the utility function you are suggesting -possibly just modifying `core.save_file`- and include that one in it.

**Comment by sami-m-g** [comment]
@cmutel - On second thought, we might want to just stick with your suggestion of having it as a different function for if user wanted to fill in defaults without/ before saving.

**Comment by cmutel** [comment]
> For generator: "pyecospold" with version set dynamically, do you mean our library version or the schema version as in 1 or 2? Can you kindly elaborate?

`pyecospold.__version__`

> `foo.timestamp`

You are right, dynamic defaults is complexity without power. Let's set these at the same time as the other dynamic attributes.

> On second thought, we might want to just stick with your suggestion of having it as a different function for if user wanted to fill in defaults without/ before saving.

Really not sure here, explicit is better than implicit but calling the same function each time can become annoying. I would +0 on having a separate function instead of overriding `save_file`, but either is fine.

**Comment by sami-m-g** [comment]
Resolved through 966f641 and 3d2b875.
